### PR TITLE
[Merged by Bors] - [cu-2gbpkxg] Add position staking

### PR DESCRIPTION
**Pull request by andor0**
## Issue
https://app.clickup.com/t/2gbpkxg

### CODE REVIEW

File: frame/staking-rewards/src/lib.rs [hunk 1]
```diff
@@ -258,13 +320,80 @@ pub mod pallet {
...
+					started_at: T::UnixTime::now().as_secs(),
+					duration: duration_preset,
+					unlock_penalty: rewards_pool.lock.unlock_penalty,
+				},
+			};
+
+			rewards_pool.total_shares += boosted_amount;
+			rewards_pool.rewards = rewards;
+
+			let position_id = StakeCount::<T>::get();
```
Comment (vimukthi-git) [inline frame/staking-rewards/src/lib.rs:390]:
```suggestion
			let position_id = StakeCount::<T>::increment();
```
?

File: frame/staking-rewards/src/lib.rs [hunk 2]
```diff
@@ -258,13 +320,80 @@ pub mod pallet {
...
 		#[transactional]
 		fn stake(
-			_who: &Self::AccountId,
-			_pool_id: &Self::RewardPoolId,
-			_amount: Self::Balance,
-			_duration_preset: DurationSeconds,
-			_keep_alive: bool,
+			who: &Self::AccountId,
+			pool_id: &Self::RewardPoolId,
+			amount: Self::Balance,
```
Comment (vimukthi-git) [inline frame/staking-rewards/src/lib.rs:325]: Need to transfer this amount to an account derived based on the pallet and pool id

File: frame/staking-rewards/src/lib.rs [hunk 3]
```diff
@@ -258,13 +320,80 @@ pub mod pallet {
...
+			duration_preset: DurationSeconds,
+			keep_alive: bool,
 		) -> Result<Self::PositionId, DispatchError> {
-			Err("Not implemented".into())
+			let mut rewards_pool =
+				RewardPools::<T>::try_get(pool_id).map_err(|_| Error::<T>::RewardsPoolNotFound)?;
+
+			let (_duration, reward_multiplier) = Self::find_nearest_duration_preset(
```
[the remark this answers is not in the document]
Comment (andor0) [reply frame/staking-rewards/src/lib.rs:332]: We have `duration_preset: DurationSeconds,` as an input of `stake` function.
How can I get `reward_multiplier` from `duration_preset`? It uses to calculate `boosted_amount`.

[the remark this answers is not in the document]
Comment (vimukthi-git) [reply frame/staking-rewards/src/lib.rs:332]: We can just `rewards_pool.lock.duration_presets[duration_preset]`

File: frame/staking-rewards/src/lib.rs [hunk 4]
```diff
@@ -81,6 +83,18 @@ pub mod pallet {
 			/// End block
 			end_block: T::BlockNumber,
 		},
+		StakeCreated {
```
[the remark this answers is not in the document]
Comment (andor0) [reply frame/staking-rewards/src/lib.rs]: Fixed.

File: frame/staking-rewards/src/lib.rs [hunk 5]
```diff
@@ -248,6 +286,30 @@ pub mod pallet {
...
+			pool_id: T::RewardPoolId,
+			amount: T::Balance,
+			duration_preset: DurationSeconds,
+		) -> DispatchResult {
+			let owner = ensure_signed(origin)?;
+			let keep_alive = true;
+			let position_id =
+				<Self as Staking>::stake(&owner, &pool_id, amount, duration_preset, keep_alive)?;
+
+			Self::deposit_event(Event::<T>::StakeCreated {
```
[the remark this answers is not in the document]
Comment (andor0) [reply frame/staking-rewards/src/lib.rs]: Done.

File: frame/staking-rewards/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (andor0) [reply frame/staking-rewards/src/lib.rs]: Fixed.

File: frame/staking-rewards/src/lib.rs [hunk 6]
```diff
@@ -258,13 +320,80 @@ pub mod pallet {
...
+					.total_rewards
+					.safe_add(&inflation)
+					.map_err(|_| ArithmeticError::Overflow)?;
+				reward.total_dilution_adjustment = reward
+					.total_dilution_adjustment
+					.safe_add(&inflation)
+					.map_err(|_| ArithmeticError::Overflow)?;
+
+				reductions
+					.try_insert(asset_id.clone(), reward.clone())
```
[the remark this answers is not in the document]
Comment (andor0) [reply frame/staking-rewards/src/lib.rs]: Fixed.

File: frame/staking-rewards/src/lib.rs [hunk 7]
```diff
@@ -258,13 +320,80 @@ pub mod pallet {
...
+			let mut rewards_pool =
+				RewardPools::<T>::try_get(pool_id).map_err(|_| Error::<T>::RewardsPoolNotFound)?;
+
+			let (_duration, reward_multiplier) = Self::find_nearest_duration_preset(
+				&rewards_pool.lock.duration_presets,
+				duration_preset,
+			)
+			.ok_or(Error::<T>::RewardConfigProblem)?;
+			let boosted_amount = reward_multiplier * amount;
+			let new_pool_shares = boosted_amount;
```
[the remark this answers is not in the document]
Comment (andor0) [reply frame/staking-rewards/src/lib.rs]: Fixed.

File: frame/staking-rewards/src/lib.rs (hunk 1 above)
Comment (andor0) [reply frame/staking-rewards/src/lib.rs]: Fixed.

File: frame/staking-rewards/src/lib.rs [hunk 8]
```diff
@@ -258,13 +313,86 @@ pub mod pallet {
...
+			let mut rewards_pool =
+				RewardPools::<T>::try_get(pool_id).map_err(|_| Error::<T>::RewardsPoolNotFound)?;
+			let reward_multiplier = rewards_pool
+				.lock
+				.duration_presets
+				.get(&duration_preset)
+				.ok_or(Error::<T>::RewardConfigProblem)?;
+
+			let boosted_amount = *reward_multiplier * amount;
+			let mut reductions: BoundedBTreeMap<
```
Comment (vimukthi-git) [inline frame/staking-rewards/src/lib.rs:331]: refer https://github.com/ComposableFi/composable/pull/1252/files#diff-88678e42185ab31dbec7ae067d5892747a99c48f339120d20435ca4cb0556c41R43

File: frame/staking-rewards/src/lib.rs [hunk 9]
```diff
@@ -258,13 +313,86 @@ pub mod pallet {
...
-			Err("Not implemented".into())
+			let mut rewards_pool =
+				RewardPools::<T>::try_get(pool_id).map_err(|_| Error::<T>::RewardsPoolNotFound)?;
+			let reward_multiplier = rewards_pool
+				.lock
+				.duration_presets
+				.get(&duration_preset)
+				.ok_or(Error::<T>::RewardConfigProblem)?;
+
+			let boosted_amount = *reward_multiplier * amount;
```
Comment (vimukthi-git) [inline frame/staking-rewards/src/lib.rs:330]: Is this safe?

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2gbpkxg Add position staking](https://app.clickup.com/t/2gbpkxg)

File: frame/staking-rewards/src/lib.rs (hunk 2 above)
Comment (andor0) [reply frame/staking-rewards/src/lib.rs:318]: Done.

File: frame/staking-rewards/src/lib.rs (hunk 9 above)
Comment (andor0) [reply frame/staking-rewards/src/lib.rs]: I replaced it with `mul_ceil`.

File: frame/staking-rewards/src/lib.rs [hunk 10]
```diff
@@ -295,4 +455,10 @@ pub mod pallet {
 			Err("Not implemented".into())
 		}
 	}
+
+	impl<T: Config> Pallet<T> {
+		pub(crate) fn pool_account_id(pool_id: &T::RewardPoolId) -> T::AccountId {
+			T::PalletId::get().into_sub_account(("po", pool_id))
```
Comment (vimukthi-git) [inline frame/staking-rewards/src/lib.rs:461]: What is `po`?

File: frame/staking-rewards/src/test/mod.rs [hunk 11]
```diff
@@ -43,6 +49,28 @@ fn test_create_reward_pool() {
...
+		let (staker, pool_id, amount, duration_preset) = (ALICE, StakingRewards::pool_count(), 100_500u32.into(), ONE_HOUR);
+
+		assert_noop!(StakingRewards::stake(Origin::signed(staker), pool_id, amount, duration_preset), crate::Error::<Test>::NotEnoughAssets);
+
+		let asset_id = StakingRewards::pools(StakingRewards::pool_count()).expect("asset_id expected").asset_id;
+		<<Test as crate::Config>::Assets as Mutate<<Test as frame_system::Config>::AccountId>>::mint_into(asset_id, &staker, amount * 2).expect("an asset minting expected");
+		assert_eq!(<<Test as crate::Config>::Assets as Inspect<<Test as frame_system::Config>::AccountId>>::balance(asset_id, &staker), amount * 2);
+
+		assert_ok!(StakingRewards::stake(Origin::signed(staker), pool_id, amount, duration_preset));
+		assert_eq!(StakingRewards::stake_count(), 1);
```
Comment (vimukthi-git) [inline frame/staking-rewards/src/test/mod.rs:68]: Would be good to verify the stake in storage with it's fields

File: frame/staking-rewards/src/lib.rs [hunk 12]
```diff
@@ -129,6 +160,11 @@ pub mod pallet {
 		/// Is used to create staked asset per `Self::RewardPoolId`
 		type CurrencyFactory: CurrencyFactory<Self::AssetId, Self::Balance>;
 
+		/// Dependency allowing this pallet to transfer funds from one account to another.
+		type Assets: Transfer<AccountIdOf<Self>, Balance = BalanceOf<Self>, AssetId = AssetIdOf<Self>>
```
Comment (vimukthi-git) [inline frame/staking-rewards/src/lib.rs:164]: Why do we need the `Mutate` bound?

Comment (andor0) [reply frame/staking-rewards/src/lib.rs]: Fixed.

File: frame/staking-rewards/src/lib.rs [hunk 13]
```diff
@@ -295,4 +450,21 @@ pub mod pallet {
...
+		}
+
+		pub(crate) fn reward_multiplier(
+			rewards_pool: &RewardPoolOf<T>,
+			duration_preset: DurationSeconds,
+		) -> Option<Perbill> {
+			rewards_pool.lock.duration_presets.get(&duration_preset).cloned()
+		}
+
+		pub(crate) fn boosted_amount(reward_multiplier: Perbill, amount: T::Balance) -> T::Balance {
```
Comment (vimukthi-git) [inline frame/staking-rewards/src/lib.rs:466]: Do we need a specific method for this?

**Review by vimukthi-git — DISMISSED** [review]
LGTM,
- Needs benchmarks
- Let's create a sub task to add prop_tests later for staking

File: frame/staking-rewards/src/lib.rs (hunk 13 above)
Comment (andor0) [reply frame/staking-rewards/src/lib.rs:466]: I created it to we can use it for tests.

**Comment by andor0** [comment]
> * Needs benchmarks

I will do it in the next task/PR. Ok?

**Comment by andor0** [comment]
> LGTM,
> 
> * Needs benchmarks

https://app.clickup.com/t/2mmqdg1

> * Let's create a sub task to add prop_tests later for staking

https://app.clickup.com/t/2mmqe99

**Comment by vimukthi-git** [comment]
@andor0 We need to do benchmarks in each PR as per convention that is agreed

**Comment by andor0** [comment]
> @andor0 We need to do benchmarks in each PR as per convention that is agreed

hmm... the pallet API can be changed during its development,
i.e. benchmarks can be updated/removed, it can happen several times,
i.e. it can take more time than needed.

ok, I will do it.

**Comment by andor0** [comment]
> @andor0 We need to do benchmarks in each PR as per convention that is agreed

I added benchmarks.

File: frame/staking-rewards/src/test/mod.rs [hunk 14]
```diff
@@ -43,6 +51,45 @@ fn test_create_reward_pool() {
 	});
 }
 
+#[test]
+fn test_stake() {
```
Comment (vimukthi-git) [inline frame/staking-rewards/src/test/mod.rs:55]: We need another test that actually has dilution protection logic kicks in

Comment (andor0) [reply frame/staking-rewards/src/test/mod.rs:55]: Could you take more details or a link to this check?

Comment (vimukthi-git) [reply frame/staking-rewards/src/test/mod.rs:55]: Make a test that has non zero dilution protection amount.

File: frame/staking-rewards/src/lib.rs (hunk 10 above)
Comment (vivekvpandya) [reply frame/staking-rewards/src/lib.rs:505]: after recent polkadot upgrade into_sub_account is renamed to into_sub_account_truncating

File: frame/staking-rewards/src/test/mod.rs (hunk 14 above)
Comment (andor0) [reply frame/staking-rewards/src/test/mod.rs:58]: I added the test for this case.

File: frame/staking-rewards/src/test/mod.rs [hunk 15]
```diff
@@ -47,6 +54,79 @@ fn test_create_reward_pool() {
...
+		}
+		rewards_pool.rewards = inner_rewards.try_into().expect("rewards expected");
+		rewards_pool.total_shares = 100u128.into();
+		RewardPools::<Test>::insert(pool_id, rewards_pool.clone());
+
+		assert_ok!(StakingRewards::stake(Origin::signed(staker), pool_id, amount, duration_preset));
+		assert_eq!(StakingRewards::stake_count(), 2);
+		let rewards_pool = StakingRewards::pools(pool_id).expect("rewards_pool expected");
+		let reward_multiplier = StakingRewards::reward_multiplier(&rewards_pool, duration_preset).expect("reward_multiplier expected");
+		let reductions = Reductions::try_from(rewards_pool.rewards.into_inner().iter().map(|(asset_id, _reward)| (*asset_id, 1005)).collect::<BTreeMap<_, _>>()).expect("reductions expected");
```
Comment (vimukthi-git) [inline frame/staking-rewards/src/test/mod.rs:110]: Could you make a comment as to the current state at this state how `1005` of reductions happen for the staker with actual numbers? I see the following
```
Pool total shares = 100
amount staked by new staker = 100,500
boosted shares = 100,500 / 100 = 1005
reduction = 100 * 1005 / 100 = 1005
```

Comment (vimukthi-git) [reply frame/staking-rewards/src/test/mod.rs:110]: I'm generally not satisfied with the readability of this test tbh

**Review by vimukthi-git — APPROVED** [review]

File: frame/staking-rewards/src/test/mod.rs [hunk 16]
```diff
@@ -47,6 +54,79 @@ fn test_create_reward_pool() {
...
+					started_at: <Test as crate::Config>::UnixTime::now(),
+					duration: duration_preset,
+					unlock_penalty: rewards_pool.lock.unlock_penalty,
+				},
+			})
+		);
+		assert_eq!(<<Test as crate::Config>::Assets as Inspect<<Test as frame_system::Config>::AccountId>>::balance(asset_id, &staker), amount * 2);
+		assert_eq!(<<Test as crate::Config>::Assets as Inspect<<Test as frame_system::Config>::AccountId>>::balance(asset_id, &StakingRewards::pool_account_id(&pool_id)), amount);
+
+		// case #2, rewards_pool.total_shares != 0 && rewards_pool.rewards[].total_rewards != 0 && inflation != 0
```
[the remark this answers is not in the document]
Comment (andor0) [reply frame/staking-rewards/src/test/mod.rs]: I splitted the test into several tests.

File: frame/staking-rewards/src/test/mod.rs (hunk 15 above)
Comment (andor0) [reply frame/staking-rewards/src/test/mod.rs]: I added an explanation of how we get inflation.

**Review by vimukthi-git — APPROVED** [review]
LGTM

**Review by vivekvpandya — APPROVED** [review]

**Pull request closed without merging**